source/extensions/filters/common/expr/evaluator.cc

```diff
@@ -138,14 +204,70 @@ BuilderInstanceSharedPtr createBuilder(Protobuf::Arena* arena) {
     throw CelException(absl::StrCat("failed to register extension regex functions: ",
                                     ext_register_status.message()));
   }
-  return std::make_shared<BuilderInstance>(std::move(builder));
+  // Register string extension functions only if enabled in configuration.
+  if (enable_string_functions) {
+    auto string_register_status =
+        cel::extensions::RegisterStringsFunctions(builder->GetRegistry(), options);
+    if (!string_register_status.ok()) {
+      throw CelException(absl::StrCat("failed to register extension string functions: ",
+                                      string_register_status.message()));
+    }
+  }
+  return builder;
 }
 
-SINGLETON_MANAGER_REGISTRATION(expression_builder);
+// Cache to store builders for different configurations.
+class BuilderCache : public Singleton::Instance, public std::enable_shared_from_this<BuilderCache> {
+public:
+  using ConfigHash = size_t;
 
-BuilderInstanceSharedConstPtr getBuilder(Server::Configuration::CommonFactoryContext& context) {
-  return context.singletonManager().getTyped<BuilderInstance>(
-      SINGLETON_MANAGER_REGISTERED_NAME(expression_builder), [] { return createBuilder(nullptr); });
+  BuilderInstanceSharedConstPtr
+  getOrCreateBuilder(OptRef<const envoy::config::core::v3::CelExpressionConfig> config) {
+    ASSERT_IS_MAIN_OR_TEST_THREAD();
+
+    ConfigHash hash = 0;
+    if (config.has_value()) {
+      // Use MessageUtil::hash for proto hashing.
+      hash = MessageUtil::hash(config.ref());
+    }
+
+    auto it = builders_.find(hash);
+    if (it != builders_.end()) {
+      auto locked_builder = it->second.lock();
+      if (locked_builder) {
+        return locked_builder;
+      }
+    }
+
+    // Create new builder with the configuration.
+    auto builder = createBuilder(config);
```

Comment:
I think this could just return `BuilderConstPtr` then? Why do you need a mutable builder?